### PR TITLE
PPS full sim specific files only (second part of the splited PR 24420)

### PR DESCRIPTION
**Pull request by mundim**
PPS full simulation code, with new SD for the tracker and timing detectors. It depends on the ongoing PR 24490 (PPSTools and new SimTransport with realistic optics parameterization from TOTEM).

@fabferro @disonjd @jan-kaspar @nminafra @setesami @forthommel @grzanka @nogima

### CODE REVIEW

**Comment by civanch** [comment]
please test

**Comment by civanch** [comment]
please test

**Comment by ianna** [comment]
please test with #24490

**Comment by mundim** [comment]
> the error related to the random number was corrected however the others still pending.
> @dilsonjd

it seems one of the problems is that the tests are issuing the global tag auto:run1_mc but in run 1 there was no PPS.

**Comment by kpedro88** [comment]
It looks like most of the test failures are due to the random engine problem. However, there was also a code rule violation:
```
Rule3 : https://raw.githubusercontent.com/cms-sw/cmssw/master/Utilities/ReleaseScripts/python/cmsCodeRules/config.py
Search for "catch(...)" statements in *.cc, *.cxx files

/SimPPS/RPDigiProducer/src/RPDisplacementGenerator.cc
[31]
```

**Comment by kpedro88** [comment]
please test

**Comment by kpedro88** [comment]
please abort

**Comment by kpedro88** [comment]
please test with #24490

**Comment by kpedro88** [comment]
Now we have the issue of missing records in the GT:
```
----- Begin Fatal Exception 21-Sep-2018 17:24:07 CEST-----------------------
An exception of category 'NoRecord' occurred while
   [0] Processing  Event run: 1 lumi: 1 event: 1 stream: 0
   [1] Running path 'digitisation_step'
   [2] Calling method for module CTPPSPixelDigiProducer/'RPixDetDigitizer'
Exception Message:
No "CTPPSPixelGainCalibrationsRcd" record found in the EventSetup.n
 Please add an ESSource or ESProducer that delivers such a record.
----- End Fatal Exception -------------------------------------------------
```

Either dummy records should be added to the old GTs (or separate ESProducers), or a modifier should be introduced so the PPS digi modules are only included in relevant workflows.

**Comment by kpedro88** [comment]
@mundim what is the status of this PR?

**Comment by Dr15Jones** [comment]
Even better would be to do 
```
if( auto rec = iSetup.tryToGet<VeryForwardMisalignedGeometryRecord>()){
   rec.get(alignments);
   ...
```
Since throwing and catching exceptions can be expensive.

**Comment by jan-kaspar** [comment]
Great, thanks @Dr15Jones !

**Comment by davidlange6** [comment]
Your PR is unmergeable. Please have a look and possibly rebase it.

**Comment by kpedro88** [comment]
@dilson there is a merge conflict. Please follow this guide to rebase: http://cms-sw.github.io/tutorial-resolve-conflicts.html

**Comment by kpedro88** [comment]
@mundim to what release did you rebase? it seems that it is not a new enough release, as the merge conflict still exists. 10_3_0_pre5 should work.

**Comment by kpedro88** [comment]
please test

**Comment by dilsonjd** [comment]
@kpedro88  the rebase is done to CMSSW_10_3_0_pre5

**Comment by ianna** [comment]
+1

**Review by kpedro88 — COMMENTED** [review]
Here is a partial review.

File: Configuration/StandardSequences/python/DigiToRaw_cff.py [hunk 1]
```diff
@@ -17,15 +17,21 @@
 from EventFilter.CastorRawToDigi.CastorDigiToRaw_cfi import *
 from EventFilter.RawDataCollector.rawDataCollector_cfi import *
 from L1Trigger.Configuration.L1TDigiToRaw_cff import *
+from EventFilter.CTPPSRawToDigi.ctppsDigiToRaw_cff import *
+
+from Configuration.ProcessModifiers.premix_stage1_cff import premix_stage1
```
[the remark this answers is not in the document]
Comment (dilsonjd) [reply Configuration/StandardSequences/python/DigiToRaw_cff.py]: Ok

File: Configuration/StandardSequences/python/Eras.py [hunk 2]
```diff
@@ -55,7 +55,8 @@ def __init__(self):
                            'trackingLowPU', 'trackingPhase1', 'ctpps_2016', 'trackingPhase2PU140','highBetaStar_2018',
                            'tracker_apv_vfp30_2016', 'pf_badHcalMitigation', 'run2_miniAOD_80XLegacy','run2_miniAOD_94XFall17', 'run2_nanoAOD_92X',
                            'run2_nanoAOD_94XMiniAODv1', 'run2_nanoAOD_94XMiniAODv2', 'run2_nanoAOD_94X2016',
-                           'hcalHardcodeConditions', 'hcalSkipPacker']
+                           'hcalHardcodeConditions', 'hcalSkipPacker',
+                           'pps_2016', 'pps_2017', 'pps_2018']
```
[the remark this answers is not in the document]
Comment (dilsonjd) [reply Configuration/StandardSequences/python/Eras.py]: The plan is to include them in Run2_2016, Run2_2017 and Run2_2018 respectively

File: EventFilter/CTPPSRawToDigi/interface/CTPPSPixelDataFormatter.h [hunk 3]
```diff
@@ -61,6 +65,8 @@ class CTPPSPixelDataFormatter {
   typedef uint32_t Word32;
   typedef uint64_t Word64;
 
+  typedef std::map<cms_uint32_t,DetDigis> Digis;
```
[the remark this answers is not in the document]
Comment (dilsonjd) [reply EventFilter/CTPPSRawToDigi/interface/CTPPSPixelDataFormatter.h]: I will check

File: EventFilter/CTPPSRawToDigi/plugins/CTPPSPixelDigiToRaw.cc [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+      // ----------member data ---------------------------
+  unsigned long eventCounter_;
+  int allDigiCounter_;
+  int allWordCounter_;
+  bool debug_;
+  std::set<unsigned int> fedIds_;
+  std::string mappingLabel_;
+  edm::ESWatcher<CTPPSPixelDAQMappingRcd> recordWatcher_;
+  edm::EDGetTokenT<edm::DetSetVector<CTPPSPixelDigi>> tCTPPSPixelDigi_;
+  std::map<std::map<const uint32_t, std::map<short unsigned int, short unsigned int> > ,  std::map<short unsigned int, short unsigned int>  > iDdet2fed_;
```
[the remark this answers is not in the document]
Comment (dilsonjd) [reply EventFilter/CTPPSRawToDigi/plugins/CTPPSPixelDigiToRaw.cc]: this triple-nested map was used to make the association between the CTPPSPixelDAQMapping and the CTPPSPixelDigi.

[the remark this answers is not in the document]
Comment (Dr15Jones) [reply EventFilter/CTPPSRawToDigi/plugins/CTPPSPixelDigiToRaw.cc]: That is not a triple-nested map. IT is a double nested map with using another map as the key!

[the remark this answers is not in the document]
Comment (kpedro88) [reply EventFilter/CTPPSRawToDigi/plugins/CTPPSPixelDigiToRaw.cc]: But the map key has a map value, so there are three levels of maps...

[the remark this answers is not in the document]
Comment (kpedro88) [reply EventFilter/CTPPSRawToDigi/plugins/CTPPSPixelDigiToRaw.cc]: @dilsonjd can you give a concrete example of how this is supposed to work? (We're not even sure what the comparison operator is when you have a map as a key...)

[the remark this answers is not in the document]
Comment (Dr15Jones) [reply EventFilter/CTPPSRawToDigi/plugins/CTPPSPixelDigiToRaw.cc]: We just looked it up, the standard does define a lexical comparison operator for map :(

[the remark this answers is not in the document]
Comment (dilsonjd) [reply EventFilter/CTPPSRawToDigi/plugins/CTPPSPixelDigiToRaw.cc]: You the use of it at EventFilter/CTPPSRawToDigi/src/CTPPSPixelDataFormatter.cc#L217 
- First there is a comparison between the information from the map with the CTPPSPixelDigi #222   
- then the other components of the map are used to build raw word. #L222

[the remark this answers is not in the document]
Comment (dilsonjd) [reply EventFilter/CTPPSRawToDigi/plugins/CTPPSPixelDigiToRaw.cc]: You can see the use of it at EventFilter/CTPPSRawToDigi/src/CTPPSPixelDataFormatter.cc#L217

- first there is a comparison between the information from the map with the CTPPSPixelDigi #222
- then the other components of the map are used to build raw word. #L222

[the remark this answers is not in the document]
Comment (makortel) [reply EventFilter/CTPPSRawToDigi/plugins/CTPPSPixelDigiToRaw.cc]: If you don't do any searching (and possibly even then), vector of pairs is much better data structure. Looking at the filling code
https://github.com/cms-sw/cmssw/blob/040f6b6d1f1b4c9a8f5d5f1f5d5322d934a01bc1/EventFilter/CTPPSRawToDigi/plugins/CTPPSPixelDigiToRaw.cc#L155-L172
I see that each of the contained maps contains exactly one element. If that is really the intended use case, I'd suggest to use `std::vector<std::tuple<uint32_t, short unsigned int, short unsigned int, short unsigned int, short unsigned int> >` (or wrapping the values to a `struct` would be even better), sorting it (by `rawId` and `rocID`), and in
https://github.com/cms-sw/cmssw/blob/040f6b6d1f1b4c9a8f5d5f1f5d5322d934a01bc1/EventFilter/CTPPSRawToDigi/src/CTPPSPixelDataFormatter.cc#L217-L245
use `std::binary_search` to look for the item (again by `rawId`, 'rocID`).

File: EventFilter/CTPPSRawToDigi/interface/CTPPSPixelDataFormatter.h [hunk 5]
```diff
@@ -69,26 +75,30 @@ class CTPPSPixelDataFormatter {
...
   int checkError(const Word32& data) const;
 
   std::string print(const Word64& word) const;
 
   const std::map<CTPPSPixelFramePosition, CTPPSPixelROCInfo> &m_Mapping;
 
+  mutable int m_DigiCounter;
+  int m_allDetDigis;
+  int m_hasDetDigis;
+  CTPPSPixelIndices theIndices;
```
Comment (perrotta) [inline EventFilter/CTPPSRawToDigi/interface/CTPPSPixelDataFormatter.h]: Maybe m_Indices, to use the same convention for the names of the class members

File: EventFilter/CTPPSRawToDigi/src/CTPPSTotemDataFormatter.cc [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+      mapIdCh[newrawId].push_back(channel); 
+    } 
+    for (auto &p : iDdet2fed) {	
+      for (auto &pf : p.first) {
+	for(auto &pId : mapIdCh) {
+	  if (pf.first == pId.first) { //compare RawIds from iDdet2fed and mapIdCh	
+	    for (int b = 0; b < 12; b++) { buf[b] = 0; bufCRC[b] = 0; } 	
+	    TotemRPDetId chipId(pf.first);
+	    for( auto &ch : pId.second ) {
+        int chInWord = ch/16;
```
Comment (perrotta) [inline EventFilter/CTPPSRawToDigi/src/CTPPSTotemDataFormatter.cc]: Indentation

File: EventFilter/CTPPSRawToDigi/src/CTPPSTotemDataFormatter.cc [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+  for (RI itFed = words.begin(); itFed != words.end(); itFed++) {
+    int fedId = itFed->first;
+    int wordsS = words.find(fedId)->second.size();
+    //due to OrbitCounter block at RawDataUnpacker 
+    words16[fedId].push_back(Word16(0)); 	  
+    words16[fedId].push_back(Word16(0)); 	 
+    //writing data in 16-bit words
+    for (int k = 0; k < wordsS; k++) {
+      for (int b = 0; b < 12; b++){
+	words16[fedId].push_back(words.find(fedId)->second.at(k)[b]);
```
Comment (perrotta) [inline EventFilter/CTPPSRawToDigi/src/CTPPSTotemDataFormatter.cc]: Indentation

File: Utilities/PPS/bin/PPSOpticsCalibration.cc [hunk 8]
```diff
@@ -0,0 +1,591 @@
...
+         ("output"          ,po::value<std::string>(&OutputFileName)->default_value("beamprofile.root"),"Output filename")
+    ;
+ 
+    po::variables_map vm;
+    po::store(po::parse_command_line(argc, argv, desc), vm);
+    po::notify(vm);    
+
+    if (vm.count("help")) { std::cout << desc << "\n"; return 1; }
+
+    //if (vm.count("Nevents")) Nevents=vm["Nevents"].as<int>();
```
Comment (perrotta) [inline Utilities/PPS/bin/PPSOpticsCalibration.cc]: Remove commented out code if not needed

File: Utilities/PPS/bin/PPSOpticsCalibration.cc [hunk 9]
```diff
@@ -0,0 +1,591 @@
+#include "Utilities/PPS/interface/PPSOpticsCalibrator.h"
```
Comment (perrotta) [inline Utilities/PPS/bin/PPSOpticsCalibration.cc]: Maybe this file should be called ```PPSOpticsCalibrator.cc```, instead?
Out of curiosity: why keeping the implementation file in the /bin subdirectory?

**Comment by davidlange6** [comment]
> On Oct 3, 2018, at 9:59 PM, Chris Jones <notifications@github.com> wrote:
> 
> @Dr15Jones commented on this pull request.
> 
> In EventFilter/CTPPSRawToDigi/plugins/CTPPSPixelDigiToRaw.cc:
> 
> > +
> +   private:
> +      void beginStream(edm::StreamID) override;
> +      void produce(edm::Event&, const edm::EventSetup&) override;
> +      void endStream() override;
> +
> +      // ----------member data ---------------------------
> +  unsigned long eventCounter_;
> +  int allDigiCounter_;
> +  int allWordCounter_;
> +  bool debug_;
> +  std::set<unsigned int> fedIds_;
> +  std::string mappingLabel_;
> +  edm::ESWatcher<CTPPSPixelDAQMappingRcd> recordWatcher_;
> +  edm::EDGetTokenT<edm::DetSetVector<CTPPSPixelDigi>> tCTPPSPixelDigi_;
> +  std::map<std::map<const uint32_t, std::map<short unsigned int, short unsigned int> > ,  std::map<short unsigned int, short unsigned int>  > iDdet2fed_;
> 
> We just looked it up, the standard does define a lexical comparison operator for map :(

still - it might be a new record for complexity of data structure proposed in cmssw?

File: EventFilter/CTPPSRawToDigi/plugins/CTPPSPixelDigiToRaw.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (dilsonjd) [reply EventFilter/CTPPSRawToDigi/plugins/CTPPSPixelDigiToRaw.cc]: @makortel I implemented the suggested struct and make a vector of it. But I did not manage to use the binary_search.  What do you thing about the implementation below? 

```
      for (auto &p : v_iDdet2fed) {
        if ((p.id == rawId) && (int(p.roc) == rocID)){
          nlink = p.fedch;
          nroc = p.rocch + 1;

          CTPPSElectronicIndex cabling = {nlink, nroc, dcol, pxid};
          cms_uint32_t word =
            (cabling.link << m_LINK_shift)
            | (cabling.roc  << m_ROC_shift)
            | (cabling.dcol << m_DCOL_shift)
            | (cabling.pxid << m_PXID_shift)
            | (it.adc() << m_ADC_shift);
          words[p.fedid].push_back(word);
          m_WordCounter++;
        } //if rocID
      } // if DAQmeeting
```

[the remark this answers is not in the document]
Comment (makortel) [reply EventFilter/CTPPSRawToDigi/plugins/CTPPSPixelDigiToRaw.cc]: @dilsonjd I'm glad to hear that the vector of a struct worked. On binary search vs. linear search, it depends how many elements the vector has. Could give an estimate how many elements this vector has (like hundreds or thousands)?

[the remark this answers is not in the document]
Comment (makortel) [reply EventFilter/CTPPSRawToDigi/plugins/CTPPSPixelDigiToRaw.cc]: Ok, with hundreds of elements the O(log n) complexity of binary search should be (much) better than the linear search.

Basically it should be sufficient to define a comparison function for the struct along
```cpp
// please consider inventing better name and/or enclosing it as a static function of the class or in some namespace
bool compare(const YourStructName& a, const YourStructName& b) {
  return a.id < b.id || (a.id == b.id && a.roc < b.roc);
}
```
use that to sort the vector (after filling it)
```cpp
std::sort(iDdet2fed.begin(), iDdet2fed.end(), compare);
```
and again in the binary search
```cpp
auto found = std::binary_search(iDdet2fed.begin(), iDdet2fed.end(), YourStructConstructor(rawId, rocID), compare);
if(found != iDdet2fed.end()) {
   <contents of the if block>
}
```

**Comment by dilsonjd** [comment]
All the suggestions regarding the FullSim were implemented.

File: EventFilter/CTPPSRawToDigi/plugins/CTPPSPixelDigiToRaw.cc [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+  for (auto const &di : *digiCollection) {
+    digiCounter += (di.data).size();
+    digis[ di.id] = di.data;
+  }
+  allDigiCounter_ += digiCounter;
+  edm::ESHandle<CTPPSPixelDAQMapping> mapping;
+  if (recordWatcher_.check( iSetup )) {
+    iSetup.get<CTPPSPixelDAQMappingRcd>().get(mapping);
+    for (const auto &p : mapping->ROCMapping)    {
+      CTPPSPixelDataFormatter::PPSPixelIndex iDdet2fed = { p.second.iD, p.second.roc, p.first.getROC(), p.first.getFEDId(), p.first.getChannelIdx() };
```
Comment (kpedro88) [inline EventFilter/CTPPSRawToDigi/plugins/CTPPSPixelDigiToRaw.cc]: if you define a constructor for `PPSPixelIndex`, you can just use `v_iDdet2fed_.emplace_back(p.second.iD, p.second.roc, p.first.getROC(), p.first.getFEDId(), p.first.getChannelIdx());` and avoid the temporary

File: EventFilter/CTPPSRawToDigi/src/CTPPSTotemDataFormatter.cc [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+            int channel = nStrip - chipPosition*nCH; //channel from DIGI		
+            uint32_t newrawId = rawId+8192*chipPosition; //8192 - distance between chipIds
+            mapIdCh[newrawId].push_back(channel); 
+        }
+        for(auto &pId : mapIdCh) {
+            PPSStripIndex myTest =  {pId.first};
+
+            // the range has always at most one element
+            auto range = std::equal_range(iDdet2fed.begin(), iDdet2fed.end(), myTest, compare);
+            std::cout << " " << range.first - iDdet2fed.begin() <<" " << range.second- iDdet2fed.begin() << '\n';
```
Comment (kpedro88) [inline EventFilter/CTPPSRawToDigi/src/CTPPSTotemDataFormatter.cc]: remove cout

File: EventFilter/CTPPSRawToDigi/src/CTPPSTotemDataFormatter.cc [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+            uint32_t newrawId = rawId+8192*chipPosition; //8192 - distance between chipIds
+            mapIdCh[newrawId].push_back(channel); 
+        }
+        for(auto &pId : mapIdCh) {
+            PPSStripIndex myTest =  {pId.first};
+
+            // the range has always at most one element
+            auto range = std::equal_range(iDdet2fed.begin(), iDdet2fed.end(), myTest, compare);
+            std::cout << " " << range.first - iDdet2fed.begin() <<" " << range.second- iDdet2fed.begin() << '\n'; 
+        if(range.first != range.second) {
```
Comment (kpedro88) [inline EventFilter/CTPPSRawToDigi/src/CTPPSTotemDataFormatter.cc]: inconsistent indentation

File: EventFilter/CTPPSRawToDigi/src/CTPPSTotemDataFormatter.cc [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+            buf[11] = (15 << 12) | (0 << 8 ) | (12 << 0) ; //build trailer as RawDataUnpacker 
+            m_WordCounter++;
+            words[fedId].push_back(buf);
+        } // range
+      } // mapIdCh
+    } //digis
+
+    typedef std::map<int, vector <std::array<uint16_t,12>> >::const_iterator RI;
+    std::map<int, vector<uint16_t > > words16;
+    for (RI itFed = words.begin(); itFed != words.end(); itFed++) {
```
Comment (kpedro88) [inline EventFilter/CTPPSRawToDigi/src/CTPPSTotemDataFormatter.cc]: range-based loop

File: EventFilter/RawDataCollector/python/rawDataCollector_cfi.py [hunk 14]
```diff
@@ -34,3 +34,10 @@
 
 from Configuration.Eras.Modifier_run3_GEM_cff import run3_GEM
 run3_GEM.toModify( rawDataCollector.RawCollectionList, func = lambda list: list.append(cms.InputTag("gemPacker")) )
+
+from Configuration.Eras.Modifier_ctpps_2016_cff import ctpps_2016
+ctpps_2016.toModify(rawDataCollector.RawCollectionList, func = lambda  list: list.extend([cms.InputTag("ctppsTotemRawData"),cms.InputTag("ctppsPixelRawData")]) )
+
+#ctpps_2016.toModify(rawDataCollector.RawCollectionList, func = lambda list:list.append(cms.InputTag("ctppsTotemRawData")) )
```
Comment (kpedro88) [inline EventFilter/RawDataCollector/python/rawDataCollector_cfi.py]: delete commented-out code

File: EventFilter/CTPPSRawToDigi/plugins/CTPPSPixelDigiToRaw.cc (hunk 10 above)
Comment (dilsonjd) [reply EventFilter/CTPPSRawToDigi/plugins/CTPPSPixelDigiToRaw.cc]: Ok

File: EventFilter/CTPPSRawToDigi/src/CTPPSTotemDataFormatter.cc (hunk 11 above)
Comment (dilsonjd) [reply EventFilter/CTPPSRawToDigi/src/CTPPSTotemDataFormatter.cc]: ok

File: EventFilter/CTPPSRawToDigi/src/CTPPSTotemDataFormatter.cc (hunk 12 above)
Comment (dilsonjd) [reply EventFilter/CTPPSRawToDigi/src/CTPPSTotemDataFormatter.cc]: ok

File: EventFilter/RawDataCollector/python/rawDataCollector_cfi.py (hunk 14 above)
Comment (dilsonjd) [reply EventFilter/RawDataCollector/python/rawDataCollector_cfi.py]: done

File: EventFilter/CTPPSRawToDigi/src/CTPPSTotemDataFormatter.cc (hunk 13 above)
Comment (dilsonjd) [reply EventFilter/CTPPSRawToDigi/src/CTPPSTotemDataFormatter.cc]: done

File: EventFilter/RawDataCollector/python/rawDataCollector_cfi.py (hunk 14 above)
Comment (dilsonjd) [reply EventFilter/RawDataCollector/python/rawDataCollector_cfi.py]: anything else? Should I push the commits again?

**Comment by kpedro88** [comment]
a few comments from before still need to be addressed:
https://github.com/cms-sw/cmssw/pull/24545#discussion_r222009817
https://github.com/cms-sw/cmssw/pull/24545#discussion_r222579332
https://github.com/cms-sw/cmssw/pull/24545#discussion_r222579492

File: Utilities/PPS/bin/PPSOpticsCalibration.cc (hunk 8 above)
Comment (dilsonjd) [reply Utilities/PPS/bin/PPSOpticsCalibration.cc]: It is part of the ongoing PR #24490

File: Utilities/PPS/bin/PPSOpticsCalibration.cc (hunk 9 above)
Comment (dilsonjd) [reply Utilities/PPS/bin/PPSOpticsCalibration.cc]: It is part of the ongoing PR #24490

File: EventFilter/CTPPSRawToDigi/plugins/CTPPSTotemDigiToRaw.cc [hunk 15]
```diff
@@ -0,0 +1,221 @@
...
+//
+// member functions
+//
+
+// ------------ method called to produce the data  ------------
+void
+CTPPSTotemDigiToRaw::produce(edm::Event& iEvent, const edm::EventSetup& iSetup)
+{
+   using namespace edm;
+  using namespace std;
```
[the remark this answers is not in the document]
Comment (dilsonjd) [reply EventFilter/CTPPSRawToDigi/plugins/CTPPSTotemDigiToRaw.cc]: Done

File: EventFilter/CTPPSRawToDigi/plugins/CTPPSTotemDigiToRaw.cc [hunk 16]
```diff
@@ -0,0 +1,221 @@
...
+  // label of the CTPPS sub-system
+  //std::string subSystemName = "TrackingStrip";
+  if (recordWatcher.check( iSetup )) {
+    iSetup.get<TotemReadoutRcd>().get(mapping);
+    for (const auto &p : mapping->VFATMapping) {
+      //get TotemVFATInfo information
+      const uint32_t pID = (p.second.symbolicID).symbolicID;
+      unsigned int phwID = p.second.hwID;
+      std::map<const uint32_t, unsigned int> mapSymb;
+      mapSymb.insert(std::pair<const uint32_t, unsigned int>(pID,phwID));
```
[the remark this answers is not in the document]
Comment (dilsonjd) [reply EventFilter/CTPPSRawToDigi/plugins/CTPPSTotemDigiToRaw.cc]: done

File: EventFilter/CTPPSRawToDigi/plugins/CTPPSTotemDigiToRaw.cc [hunk 17]
```diff
@@ -0,0 +1,221 @@
...
+
+  edm::Handle< edm::DetSetVector<TotemRPDigi> > digiCollection;
+  iEvent.getByToken( tTotemRPDigi_, digiCollection);
+
+  CTPPSTotemDataFormatter::RawData rawdata;
+  CTPPSTotemDataFormatter::Digis digis;
+
+  int digiCounter = 0;
+  typedef vector< edm::DetSet<TotemRPDigi> >::const_iterator DI;
+  for (DI di=digiCollection->begin(); di != digiCollection->end(); di++) {
```
[the remark this answers is not in the document]
Comment (dilsonjd) [reply EventFilter/CTPPSRawToDigi/plugins/CTPPSTotemDigiToRaw.cc]: done

File: EventFilter/CTPPSRawToDigi/plugins/CTPPSPixelDigiToRaw.cc [hunk 18]
```diff
@@ -0,0 +1,220 @@
...
+  // start the counters
+  eventCounter_ = 0;
+  allDigiCounter_ = 0;
+  allWordCounter_ = 0; 
+}
+
+
+CTPPSPixelDigiToRaw::~CTPPSPixelDigiToRaw()
+{
+   edm::LogInfo("CTPPSPixelDigiToRaw")  << " CTPPSPixelDigiToRaw destructor!";
```
[the remark this answers is not in the document]
Comment (dilsonjd) [reply EventFilter/CTPPSRawToDigi/plugins/CTPPSPixelDigiToRaw.cc]: done

File: EventFilter/CTPPSRawToDigi/plugins/CTPPSPixelDigiToRaw.cc [hunk 19]
```diff
@@ -0,0 +1,220 @@
...
+    iSetup.get<CTPPSPixelDAQMappingRcd>().get(mapping);
+    for (const auto &p : mapping->ROCMapping)    {
+        const uint32_t piD = p.second.iD;
+        short unsigned int pROC   = p.second.roc;
+        short unsigned int pFediD = p.first.getFEDId();
+        short unsigned int pFedcH = p.first.getChannelIdx();
+        short unsigned int pROCcH = p.first.getROC();
+
+        std::map<short unsigned int,short unsigned int> mapROCIdCh;
+        mapROCIdCh.insert(std::pair<short unsigned int,short unsigned int>(pROC,pROCcH));
```
[the remark this answers is not in the document]
Comment (dilsonjd) [reply EventFilter/CTPPSRawToDigi/plugins/CTPPSPixelDigiToRaw.cc]: done

File: EventFilter/CTPPSRawToDigi/plugins/CTPPSPixelDigiToRaw.cc [hunk 20]
```diff
@@ -0,0 +1,220 @@
...
+
+        std::map<const uint32_t, std::map<short unsigned int, short unsigned int> > mapDetRoc;
+        mapDetRoc.insert(std::pair<const uint32_t, std::map<short unsigned int,short unsigned int> >(piD,mapROCIdCh));
+  std::map<short unsigned int,short unsigned int> mapFedIdCh;
```
[the remark this answers is not in the document]
Comment (dilsonjd) [reply EventFilter/CTPPSRawToDigi/plugins/CTPPSPixelDigiToRaw.cc]: done

File: EventFilter/CTPPSRawToDigi/plugins/CTPPSPixelDigiToRaw.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (dilsonjd) [reply EventFilter/CTPPSRawToDigi/plugins/CTPPSPixelDigiToRaw.cc]: done

File: EventFilter/CTPPSRawToDigi/interface/CTPPSTotemDataFormatter.h [hunk 21]
```diff
@@ -0,0 +1,62 @@
...
+    typedef uint16_t Word16;
+    typedef uint32_t Word32;
+    typedef uint64_t Word64;
+
+    mutable int theWordCounter;
+    mutable int theDigiCounter;
+
+  public:
+
+    typedef std::map<int, FEDRawData> RawData;
```
[the remark this answers is not in the document]
Comment (dilsonjd) [reply EventFilter/CTPPSRawToDigi/interface/CTPPSTotemDataFormatter.h]: done

**Comment by perrotta** [comment]
@cmsbuild please test

**Comment by perrotta** [comment]
I've just launched the tests for this PR.

However, I see that (for example) Utilities/PPS/bin/PPSOpticsCalibration.cc is called (more correctly) Utilities/PPS/bin/PPSOpticsCalibrator.cc in #24490. Clearly the two PRs are not kept synchronized (as they should): therefore, they must at least follow some order, and they must be evaluated and merged by strictly following that order.

@mundim please provide us with a map that explains:
- Which are the daughter PR of the original #24420? Only this one and #24490
- In which order they should be evaluated, and therefore merged?

Once any of them is merged, the other one(s) should get rebased: evaluating those pull requests before they being rebased is a waste of time, since it is evident that they are not kept synchronized and unavoidably they will eventually become different.

Please provide a clear plan as a comment (and also in the description part at the top) in this pull request, in #24490, and in all other possible affected PR: I cannot  continue the review till then.

**Comment by kpedro88** [comment]
The build errors are:
```
/build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_4_X_2018-10-14-2300/src/SimPPS/PPSPixelDigiProducer/interface/PPSPixelDigiProducer.h:69:20: warning: 'CTPPSPixelDigiProducer::beginRun' hides overloaded virtual function [-Woverloaded-virtual]
/build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_4_X_2018-10-14-2300/src/SimPPS/RPDigiProducer/interface/RPDigiProducer.h:52:20: warning: 'RPDigiProducer::beginRun' hides overloaded virtual function [-Woverloaded-virtual]
/build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_4_X_2018-10-14-2300/src/SimTransport/PPSProtonTransport/src/HectorTransport.cc:96:18: warning: explicitly assigning value of variable of type 'double' to itself [-Wself-assign]
```

One of these seems to be the same as a warning seen in #24490.

There is also a code rule violation:
```
Rule3 : https://raw.githubusercontent.com/cms-sw/cmssw/master/Utilities/ReleaseScripts/python/cmsCodeRules/config.py
Search for "catch(...)" statements in *.cc, *.cxx files

/SimPPS/RPDigiProducer/src/RPDisplacementGenerator.cc
[31]
```

**Comment by mundim** [comment]
> I've just launched the tests for this PR.
> 
> However, I see that (for example) Utilities/PPS/bin/PPSOpticsCalibration.cc is called (more correctly) Utilities/PPS/bin/PPSOpticsCalibrator.cc in #24490. Clearly the two PRs are not kept synchronized (as they should): therefore, they must at least follow some order, and they must be evaluated and merged by strictly following that order.
> 
> @mundim please provide us with a map that explains:
> 
>     * Which are the daughter PR of the original #24420? Only this one and #24490
Yes.

> 
>     * In which order they should be evaluated, and therefore merged?

The PR 24490 should go first.

> 
> Once any of them is merged, the other one(s) should get rebased: evaluating those pull requests before they being rebased is a waste of time, since it is evident that they are not kept synchronized and unavoidably they will eventually become different.
> 
> Please provide a clear plan as a comment (and also in the description part at the top) in this pull request, in #24490, and in all other possible affected PR: I cannot continue the review till then.

In summary: This PR depends on the PR 24490, so 24490 needs to go first. Then, this one should be rebased.  These 2 PRs (24490 and 24545) are the only PR affected.

**Comment by davidlange6** [comment]
Your PR is unmergeable. Please have a look and possibly rebase it.

**Comment by dilsonjd** [comment]
There is already a new version but as this PR depends on the PR24490 we were waiting for his approval. Do you thing that it is a good strategy? 
With which release the rebase must be done?

**Comment by perrotta** [comment]
@dilsonjd : as soon as there will be an IB which includes #24490, then rebase on that IB (or on an IB newer than that)

**Comment by perrotta** [comment]
Periodic check: is there anything new for this PR coming?

**Comment by dilsonjd** [comment]
Still waiting for an IB which includes #25355 (a clean version of #24490).

**Comment by fabiocos** [comment]
I had some comments in #25355 that I would like to see addressed here

@dilsonjd this PR needs to be rebased at this point

**Comment by civanch** [comment]
@mundim , I would suggest to close this PR and start a new one on top of merged PR #25355. The new IB will be available tomorrow. 208 files in the current PR is too much, this would be good to split, I would have 1st part with only new packages alone, in the next implement interaction with the rest of CMSSW. Fimnally we need implement comments from David and Fabio everywhere, provide unit test and/or test simulations, introduce G4Region "PPS".

**Comment by mundim** [comment]
As recommended above, closing this PR, waiting for the approval of PR #25355, when a PR for these packages will be resubmitted.

**Pull request closed without merging**